tsdb/engine/tsm1/file_store_test.go

```diff
@@ -2786,6 +2787,151 @@ func TestFileStore_CreateSnapshot(t *testing.T) {
 	}
 }
 
+func TestFileStore_ReaderBlocking(t *testing.T) {
+	dir := t.TempDir()
+
+	// Create 3 TSM files...
+	data := []keyValues{
+		keyValues{"cpu", []tsm1.Value{tsm1.NewValue(0, 1.0)}},
+		keyValues{"cpu", []tsm1.Value{tsm1.NewValue(1, 2.0)}},
+		keyValues{"mem", []tsm1.Value{tsm1.NewValue(0, 1.0)}},
+	}
+
+	files, err := newFileDir(dir, data...)
+	if err != nil {
+		fatal(t, "creating test files", err)
+	}
+
+	fs := tsm1.NewFileStore(dir)
+	if err := fs.Open(); err != nil {
+		fatal(t, "opening file store", err)
+	}
+	defer fs.Close()
+
+	fsInUse := func() bool {
+		t.Helper()
+		require.NoError(t, fs.SetNewReadersBlocked(true))
+		defer fs.SetNewReadersBlocked(false)
+		inUse, err := fs.InUse()
+		require.NoError(t, err)
+		return inUse
+	}
+
+	checkUnblocked := func() {
+		t.Helper()
+
+		require.False(t, fsInUse())
+		_, err := fs.InUse()
+
+		var applyCount atomic.Uint32
+		err = fs.Apply(func(r tsm1.TSMFile) error {
+			applyCount.Add(1)
+			return nil
+		})
+		require.NoError(t, err)
+		require.Equal(t, uint32(len(files)), applyCount.Load(), "Apply should be called for all files")
+
+		snap, err := fs.CreateSnapshot()
+		require.NoError(t, err)
+		require.NotEmpty(t, snap)
+
+		buf := make([]tsm1.FloatValue, 1000)
+		c := fs.KeyCursor(context.Background(), []byte("cpu"), 0, true)
+		defer func() {
+			if c != nil {
+				c.Close()
```

Comment:
can `Close` fail?

Reply:
No. `tsm1.KeyCursor.Close` does not have a return value.